### ISSUE TITLE
Invalid eta expansion in notation pretty-printer

### ISSUE THREAD

**Issue by samuelgruetter** [coq/coq#15221]
Continuation-passing style definitions can be cumbersome to read:

```coq
Definition foo{A}(a b: nat)(k: nat -> A): A :=
  k (a + b).

Definition bar{A}(a b: nat)(k: nat -> A): A :=
  k (a - b).

Definition chain(x y: nat): Prop :=
  foo x y (fun f =>
    bar x y (fun b =>
      f = b)).
```

A notation could make it nicer:

```coq
Notation "'let/c' x := r 'in' b" := (r (fun x => b))
  (x binder, at level 200, right associativity,
   format "'[hv' 'let/c'  x  :=  r  'in'  '//' b ']'").

Definition chain'(x y: nat): Prop :=
  let/c f := foo x y in
  let/c b := bar x y in
  f = b.

Print chain'.
```

But the notation gets printed as follows:

```
chain' =
fun x y : nat =>
(let/c x := @foo in
 let/c x0 := Prop in
 x x0) x y
  (fun f : nat => (let/c x := @bar in
                   let/c x0 := Prop in
                   x x0) x y (fun b : nat => f = b))
     : nat -> nat -> Prop
```

It seems that the notation pretty-printer over-eagerly inserts eta expansions in places where I don't want them to be inserted.

Would it be possible to add an option to disable eta expansion during notation pretty-printing on  a per-notation basis?

For instance, using a `no eta expansion` option like this?

```coq
Notation "'let/c' x := r 'in' b" := (r (fun x => b))
  (x binder, at level 200, right associativity,
   no eta expansion,
   format "'[hv' 'let/c'  x  :=  r  'in'  '//' b ']'").
```

Note: For people looking for workarounds, there are various ways of using a marker so that the notation is only printed when there's a marker. The right-hand side of the notation could be `(app r (fun x => b))` (where `Definition app {A B} (g: A -> B) x := g x.`), or `(let K := (fun x => b) in r K)`, or `((fun K => r K) (fun x => b))`, but these workarounds all have the problem that if the markers are simplified away, the notation will not be used for printing any more, and they also change the source terms, so tactics doing a `match` might stop working.

/cc @herbelin

**Comment by SkySkimmer** [comment]
This is a bug, eg `let/c x0 := Prop in _` is `Prop (fun x0 => _)` which is just wrong.

**Comment by samuelgruetter** [comment]
Good point, so this should actually be a bug report, with the following minimal example:

```coq
Notation "'let/c' x := r 'in' b" := (r (fun x => b))
  (x binder, at level 200, right associativity).

Check id 0.
```

prints

```
(let/c x := id in let/c x0 := nat in x x0) 0
     : nat
```

but passing this to `Check` results in a type error (`nat` on the rhs of `let/c` is nonsense, there can only be functions that take a continuation).

**Comment by herbelin** [comment]
Thanks for reporting. I had missed the report. Now fixed in #17841.
